lib/propolis/src/tasks.rs

```diff
@@ -0,0 +1,378 @@
+        if let Some(w) = self.waker_ctrl.as_ref() {
+            w.wake_by_ref();
+        }
+    }
+}
+struct Inner {
+    /// The `Control` of a task can be manipulated from both sync and async
+    /// contexts.  Care should be taken not to otherwise block while holding the
+    /// mutex guard, as it could cause an undue stall for any async pollers.
+    ctrl: Mutex<Control>,
+    cv: Condvar,
+    notify_fn: Option<Box<NotifyFn>>,
+}
+impl Inner {
+    fn request_hold<'a>(
+        &'a self,
+        mut guard: MutexGuard<'a, Control>,
+        ctrl_waker: Option<Waker>,
+    ) -> MutexGuard<'a, Control> {
+        guard.should_hold = true;
+        if let Some(waker) = ctrl_waker {
+            guard.waker_ctrl = Some(waker);
+        }
+        guard.wake_task();
+        self.cv.notify_all();
+        self.notify_task(guard)
+    }
+    fn unrequest_hold(&self, guard: &mut MutexGuard<Control>) {
+        guard.should_hold = false;
+        guard.waker_ctrl = None;
+        // If the task was not held already, and has not been requested to exit,
+        // then removing the hold request counts as a state of no pending events
+        let has_event = guard.is_held || guard.should_exit;
+        if has_event {
+            guard.wake_task();
+            self.cv.notify_all();
+        }
+    }
+    fn request_exit(&self, mut guard: MutexGuard<Control>) {
+        guard.should_exit = true;
+        guard.wake_task();
+        self.cv.notify_all();
+        let _guard = self.notify_task(guard);
+    }
+
+    /// Call notifier function (if one exists) for task.
+    ///
+    /// This notification may synchronously access resources which would
+    /// otherwise be exclusive of the `MutexGuard` held on the [`Control`], so
+    /// the said guard _must_ be dropped while calling the notifier.
+    fn notify_task<'a>(
+        &'a self,
+        guard: MutexGuard<'a, Control>,
+    ) -> MutexGuard<'a, Control> {
+        match self.notify_fn.as_ref() {
+            // Task notification only required if it is not already held
+            Some(notify) if !guard.is_held => {
+                drop(guard);
+                notify();
+                self.ctrl.lock().unwrap()
+            }
+            _ => guard,
+        }
+    }
+
+    fn set_held(
+        &self,
+        guard: &mut MutexGuard<Control>,
+        task_waker: Option<Waker>,
+    ) {
+        guard.is_held = true;
+        if let Some(waker) = task_waker {
+            guard.waker_task = Some(waker);
+        }
+        // Notify ctrl of changed state
+        guard.wake_ctrl();
+        self.cv.notify_all();
+    }
+    fn set_unheld(&self, guard: &mut MutexGuard<Control>) {
+        guard.is_held = false;
+        guard.waker_task = None;
+        // Notify ctrl of changed state
+        guard.wake_ctrl();
+        self.cv.notify_all();
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum Event {
+    Hold,
+    Exit,
+}
+
+pub struct TaskHdl(Arc<Inner>);
```

Comment:
Add some docs on `TaskHdl` vs `TaskCtrl` it's not clear at first glance the difference between the two especially as they have methods with the same names.

The next part shouldn't hold up the PR but is more for possible future changes: Is there any reason we can't just rename `Inner` to `Task` and just let the consumers use `Arc<Task>` or `Weak<Task>` as appropriate? To keep the same ergonomics we could keep all the current `TaskHdl` methods as-is and just change the self type they take to require the `Arc`. Unfortunately not so straightforward for `TaskCtrl` as `Weak` can't be used as a receiver type today. But either the upgrade could be pushed to the caller or just use static methods that take a `Weak` ref.

Reply:
> Add some docs on TaskHdl vs TaskCtrl it's not clear at first glance the difference between the two especially as they have methods with the same names.

Will do.

>  Is there any reason we can't just rename Inner to Task and just let the consumers use Arc<Task> or Weak<Task> as appropriate?

One aspect of the current implementation (and it wasn't always this way) is that there is only one "copy" each of the `TaskHdl` and `TaskCtrl`.  This simplifies storage of the Waker for the futures which we can initiate from either side.  It's part of the reason why the receivers for those future methods are `&mut` - so there can only be one at a time.  If we move to a naked `Arc`/`Weak`, then we lose that convenience.

Reply:
> Add some docs on TaskHdl vs TaskCtrl

I hope it's a little more clear with the added docs?

Reply:
Yes, thank you!